### ISSUE TITLE
liu attention

### ISSUE THREAD

**Issue by bringtree** [bringtree/question_embedding#146]
代码全局搜softmax
![image](https://user-images.githubusercontent.com/20106478/47866463-c8e8c400-de39-11e8-9291-a40d3d389d9f.png)
定位在这里
关注s的计算
![image](https://user-images.githubusercontent.com/20106478/47866481-d4d48600-de39-11e8-8472-f7e7de62ecc2.png)
```python3
v[i] * tf.tanh(hidden_features[i] + y)
```
跟着hidden_features的计算能看到v参数的shape
![image](https://user-images.githubusercontent.com/20106478/47866551-0d745f80-de3a-11e8-9d8a-920c64738bca.png)
出现了K
先看看做了啥，出来了[1,1,attn_size, attention_vec_size]
继续往上翻
看到
```python3
    attention_vec_size = attn_size  # Size of query vectors for attention.
```
ok，知道卷积核的样子了，大概能猜到是想做啥了
```python3
tf.nn.conv2d(hidden, k, [1, 1, 1, 1], "SAME")
```
hidden.shape = [batch_size,length of sentence,1,lstm_hidden*2双向]
>猜测:这种情况就是在做个w*h呀 所有h公用w的计算吧?
验证下：
去看下hidden 是啥
发现他叫做attention_states
![image](https://user-images.githubusercontent.com/20106478/47866808-be7afa00-de3a-11e8-837a-4c192afd5fa1.png)
又回到这
![image](https://user-images.githubusercontent.com/20106478/47866832-ccc91600-de3a-11e8-9fcb-0cc8e3f8f6d1.png)
![image](https://user-images.githubusercontent.com/20106478/47866896-f8e49700-de3a-11e8-9f09-be509c3dcfb7.png)
ok 猜想没错。是输出h。
那么公式前半部分解决hidden_features[i] 是一个 w*h,乘完不改变维度。只做空间变化。另外v是一个和h同维度的东西。是会把后面的向量重新转成一个数字。
```
v[i] * tf.tanh(hidden_features[i] + y)
```
接着挖掘y是啥
![image](https://user-images.githubusercontent.com/20106478/47867009-48c35e00-de3b-11e8-9b02-6adeceaf4319.png)
![image](https://user-images.githubusercontent.com/20106478/47867027-511b9900-de3b-11e8-9edf-cf7f4df41b51.png)
恩 公式在上面不多说，直接去找query是啥。
![image](https://user-images.githubusercontent.com/20106478/47867086-7ad4c000-de3b-11e8-8ac6-a7a66bc0d4af.png)

![image](https://user-images.githubusercontent.com/20106478/47867099-82946480-de3b-11e8-96e4-5862270d8baa.png)
![image](https://user-images.githubusercontent.com/20106478/47867111-8aec9f80-de3b-11e8-96e9-e5ea1fbfabde.png)
![image](https://user-images.githubusercontent.com/20106478/47867127-96d86180-de3b-11e8-976f-e88267ffabd4.png)
![image](https://user-images.githubusercontent.com/20106478/47867190-c7b89680-de3b-11e8-9b47-b7266f545469.png)
哦 y 是最后输出的拼接
不过这样维度对不上
前面是有 50个h，后面缺只有1个h。
做加法法。可以推测 做了广播
好了到这里就结束了
公式 \aplha_n = softmax(v*tanh(w*h_n+u*h_n))
